File: apps/nekbone/refactored/unit_tests/pnormj/nekbone_pnormj.c
#ifndef UNIT_TEST_NEKBONE_JACG_C
#define UNIT_TEST_NEKBONE_JACG_C

#include <stdio.h>
#include "blas1.h"
#include "polybasis.h"

#define Err_t int
#define IFEB if(err) break

int main(int argc, char * argv[])
{
    FILE * fout = 0;

    Err_t err=0;
    while(!err){
        fout = fopen("./z_ccode.out", "w");
        if(!fout) {err=__LINE__; IFEB;}

        double alpha, beta;
        unsigned int pDOF, firstdegree, lastdegree;

        alpha = 0; // This is how it is use throughout NEKbone.
        beta  = 0; // This is how it is use throughout NEKbone.

        // If any of these values changes, make also the change in nekbone_pnormj.f90
        firstdegree = 2;
        lastdegree  = 25;

        const unsigned int N = 84; //Largest value allowed in Nekbone's speclib.f::ZWGJ
        if(lastdegree >= N){
            printf("ERROR: test_jacg: lastdegree must be smaller than 100.\n");
            err=__LINE__; IFEB;
        }

        for(pDOF=firstdegree; pDOF<=lastdegree; ++pDOF){
            double nj = nbb_pnormj(pDOF, alpha, beta);
            fprintf(fout,"%10u %23.14E\n", pDOF, nj);
        }

        if(fout){
            fclose(fout); fout=0;
        }

        break; //while(!err){
    }

    if(fout){
        fclose(fout); fout=0;
    }

    if(err){
        printf("ERROR: %lu\n", (unsigned long)err );
        return 1;
    }
    return 0;
}

#endif // UNIT_TEST_NEKBONE_JACG_C

File: apps/nekbone/refactored/unit_tests/pnormj/nekbone_pnormj.f90
!-----------------------------------------------------------------------
! The code for the procedure "PNORMJ" was taken from
!   nekbone-2.3.4/src/speclib.f
!-----------------------------------------------------------------------
      program test_pnormj

      implicit none

      INTERFACE
        FUNCTION PNORMJ(N,ALPHA,BETA)
            INTEGER N
            REAL*8 ALPHA, BETA
        END FUNCTION PNORMJ
      END INTERFACE

      REAL*8 alpha, beta

      integer pDOF, N
      integer firstdegree, lastdegree

      REAL*8 nj

      alpha = 0. ! This is how it is use throughout NEKbone
      beta  = 0. ! This is how it is use throughout NEKbone

      ! If any of these values changes, make also the change in nekbone_zwgjd.c
      firstdegree = 2
      lastdegree  = 25

      N = 84 ! Largest value allowed in Nekbone's speclib.f::ZWGJ
      if( lastdegree > N) then
        write(*,*) "ERROR: test_zwgjd: lastdegree must be smaller than 84"
        call exit(1)
      endif

      open(unit=10,file='z_fortran.out')

      do pDOF=firstdegree,lastdegree
        nj = PNORMJ(pDOF,alpha,beta)
        write(10,'(1I10,1ES23.14)') pDOF, nj
      enddo

      close(10)

      call exit(0)
      end

!--------------------------------------------------------------------
     REAL*8  FUNCTION GAMMAF (X)
      IMPLICIT REAL*8  (A-H,O-Z)
      REAL*8  X
      ZERO = 0.0
      HALF = 0.5
      ONE  = 1.0
      TWO  = 2.0
      FOUR = 4.0
      PI   = FOUR*ATAN(ONE)
      GAMMAF = ONE
      IF (X.EQ.-HALF) GAMMAF = -TWO*SQRT(PI)
      IF (X.EQ. HALF) GAMMAF =  SQRT(PI)
      IF (X.EQ. ONE ) GAMMAF =  ONE
      IF (X.EQ. TWO ) GAMMAF =  ONE
      IF (X.EQ. 1.5  ) GAMMAF =  SQRT(PI)/2.
      IF (X.EQ. 2.5) GAMMAF =  1.5*SQRT(PI)/2.
      IF (X.EQ. 3.5) GAMMAF =  0.5*(2.5*(1.5*SQRT(PI)))
      IF (X.EQ. 3. ) GAMMAF =  2.
      IF (X.EQ. 4. ) GAMMAF = 6.
      IF (X.EQ. 5. ) GAMMAF = 24.
      IF (X.EQ. 6. ) GAMMAF = 120.
      RETURN
      END

!--------------------------------------------------------------------
      REAL*8  FUNCTION PNORMJ (N,ALPHA,BETA)
      IMPLICIT REAL*8  (A-H,O-Z)
      REAL*8  ALPHA,BETA
      ONE   = 1.
      TWO   = 2.
      DN    = ((N))
      CONST = ALPHA+BETA+ONE
      IF (N.LE.1) THEN
         PROD   = GAMMAF(DN+ALPHA)*GAMMAF(DN+BETA)
         PROD   = PROD/(GAMMAF(DN)*GAMMAF(DN+ALPHA+BETA))
         PNORMJ = PROD * TWO**CONST/(TWO*DN+CONST)
         RETURN
      ENDIF
      PROD  = GAMMAF(ALPHA+ONE)*GAMMAF(BETA+ONE)
      PROD  = PROD/(TWO*(ONE+CONST)*GAMMAF(CONST+ONE))
      PROD  = PROD*(ONE+ALPHA)*(TWO+ALPHA)
      PROD  = PROD*(ONE+BETA)*(TWO+BETA)
      DO 100 I=3,N
         DINDX = ((I))
         FRAC  = (DINDX+ALPHA)*(DINDX+BETA)/(DINDX*(DINDX+ALPHA+BETA))
         PROD  = PROD*FRAC
 100  CONTINUE
      PNORMJ = PROD * TWO**CONST/(TWO*DN+CONST)
      RETURN
      END
